opencog/server/BaseServer.h

```diff
@@ -65,8 +65,41 @@ class BaseServer
     virtual ~BaseServer(void);
 }; // class
 
-// singleton instance (following meyer's design pattern)
+// Singleton instance (following meyer's design pattern) NOTE: There are
+// problems with the way this is being used, in some cases, as it is confusing
+// that there are examples of both stack based and factory function singleton
+// server creation in the tests and examples. For example, I saw some code
+// that did this:
+//
+// 	  void testAddPunishmentPredicate() {
+//        server(SpaceTimeCogServer::createInstance);  << creates one instance
+//        SpaceTimeCogServer srv;                      << creates another
+//        AtomSpace& atomSpace = srv.getAtomSpace();
+//
```

Comment:
The whole design of the CogServer, SpaceTimeCogServer, etc., needs to be
radically changed, as we've discussed before

However, this is a big job in itself, and the fixing of the python
interface shouldn't wait for this, obviously...

On Sat, Feb 28, 2015 at 12:51 PM, Linas Vepštas <notifications@github.com>
wrote:

> In opencog/server/BaseServer.h
> <https://github.com/opencog/opencog/pull/1325#discussion_r25554388>:
>
> > @@ -65,8 +65,41 @@ class BaseServer
> >      virtual ~BaseServer(void);
> >  }; // class
> >
> > -// singleton instance (following meyer's design pattern)
> > +// Singleton instance (following meyer's design pattern) NOTE: There are
> > +// problems with the way this is being used, in some cases, as it is confusing
> > +// that there are examples of both stack based and factory function singleton
> > +// server creation in the tests and examples. For example, I saw some code
> > +// that did this:
> > +//
> > +// 	  void testAddPunishmentPredicate() {
> > +//        server(SpaceTimeCogServer::createInstance);  << creates one instance
> > +//        SpaceTimeCogServer srv;                      << creates another
> > +//        AtomSpace& atomSpace = srv.getAtomSpace();
> > +//
>
> Yes, all this needs fixing. Its not at all clear what the correct fix
> should be. The space server may need to be replaced by or integrated with a
> ROS-based space server, so that spatial and motion-planning information can
> be shared with robot platforms. This is still wildly unclear to me.
>
> The Embodiment server inherits from the CogServer, but I don't know of any
> reason that it should. I don't see how it derives any benefit from doing
> so. It seems like some old arbitrary design decision. Again, its not clear
> to me what the correct fix should be.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/pull/1325/files#r25554388>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw